### ISSUE TITLE
windows-mingw build problems

### ISSUE THREAD

**Issue by dopefishh** [wolfSSL/wolfMQTT#177]
Thank you anyway very much for the library, we are starting to use it on esp8266 devices and linux machines for the [mTask](https://gitlab.science.ru.nl/mlubbers/mtask) system and it's working promisingly well there. However, when trying to port the client to windows I encountered some problems:

When I try to cross-compile wolfMQTT for windows on my linux box I a mingw environment encounter several problems. I've also tried this on a native mingw windows system and have the same issues.

```
./autogen.sh
./configure --disable-tls --host=x86_64-pc-mingw32 CC=x86_64-w64-mingw32-gcc
make
```

- When I disable tls (`./configure --disable-tls`) it still tries to include wolfssl libraries. This is because the `ENABLE_MQTT_TLS` definition is set even in mingw environments in `vs_types.h`. I thought `vs_types.h` was only required for visual studio. If I remove the definition from `vs_types.h` it works.
- Then I proceed and get several attribute errors:

  ```
  src/mqtt_packet.c:3404:1: error: visibility attribute not supported in this configuration; ignored [-Werror=attributes]
  ```

  They can be turned into warnings by adding `-Wno-attributes` to the CFLAGS.
- Then I proceed to compile and encounter the following error:

```
make
...
  CC       src/libwolfmqtt_la-mqtt_client.lo
  CC       src/libwolfmqtt_la-mqtt_packet.lo
examples/mqttnet.c: In function ‘NetWrite’:
examples/mqttnet.c:814:26: error: comparison is always false due to limited range of
data type [-Werror=type-limits]
             if (so_error == EWOULDBLOCK || so_error == EAGAIN) {
                          ^~
examples/mqttnet.c: In function ‘NetRead_ex’:
examples/mqttnet.c:962:26: error: comparison is always false due to limited range of
data type [-Werror=type-limits]
             if (so_error == EWOULDBLOCK || so_error == EAGAIN) {
                          ^~
cc1: all warnings being treated as errors
  CC       src/libwolfmqtt_la-mqtt_socket.lo
```

TLDR:

- What is the meaning of `vs_types.h` and is it required to modify it in cross-compile mingw environments as well?
- What do the attributes do and why are they not supported in cross-compile mingw environments?
- I found some errors in the `mqttnet.c` example.

**Comment by dgarske** [comment]
Hi @dopefishh , 

I've pushed some fixes here:
https://github.com/wolfSSL/wolfMQTT/pull/178

For the mingw cross-compile you might try the following:

```
./configure --host=i686 CC=i686-w64-mingw32-gcc LD=i686-w64-mingw32-ld CFLAGS="-DWIN32 -DMINGW -D_WIN32_WINNT=0x0600" LIBS="-lws2_32" --disable-tls
```

The visibility errors are resolved by adding `__MINGW32__` check in `wolfmqtt/visibility.h`.
The non-blocking error is resolved by not using/letting the wolfmqtt/vs_settings.h build options be included for autoconf cross-compile.

Let me know if you have any other issues.

Thanks,
David Garske, wolfSSL

**Comment by dopefishh** [comment]
Thanks, it works now. Also, it seems to works on 64 bit as well.
